proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,657 @@
+   the global scope but not the others.
+
+3. When a function ``f`` needs a helper ``h``, we can declare ``h`` in a
+   ``where`` clause. However, suppose both ``f`` and ``g`` need ``h``. Now,
+   ``h`` must be declared at the top level, meaning all the rest of the
+   definitions in the module can see ``h``, even if ``h`` is really very
+   specific to ``f`` and ``g``.
+
+   With this proposal, we can model this situation nicely::
+
+     import module _ (f, g) where
+       f :: ...
+       f = ...
+
+       g :: ...
+       g = ...
+
+       h :: ...
+       h = ...
+
+   Unfortunately, there is no way for ``h`` to access arguments passed to
+   ``f`` or ``g`` without declaring these as arguments to ``h`` and passing
+   them explicitly. Fixing this was the subject of `#40`_; see `Future Work`_.
+
+4. If a function or group of functions needs to work with a datatype or class
+   locally, there is no way to do this without polluting the namespace of the
+   entire file.
+
+   With this proposal, we can do this easily::
+
+     import module _ (f) where
+       f :: ...
+       f = ...
+
+       data SpecialDataType = ...
+
+       class LocalClass a b c where ...
+
+5. When we expect users to import our library qualified, we have no way of
+   signaling this beyond documentation; we also have no way of making it easy
+   to import the module qualified correctly.
+
+   With this proposal, we can do this easily::
+
+     -- top of file:
+     module Data.Set ( Set, qualified module Set ) where
+
+       import module Set ( Set )
+       module Set ( Set, fromList ) where
+         data Set = ...
+         fromList = ...
+
+   An importer who days ``import Data.Set`` will get access to ``Set`` (the
+   type) and ``Set.fromList``, the function. The fact that the module and type
+   have the same name is inconsequential here, but it is permitted.
+
+6. Suppose our library exports class ``C`` with method ``meth``. Our users
+   will frequently import ``C(..)`` and get ``meth`` in scope. Now I wish to
+   refactor ``C`` without changing my users' import behavior. There is no way
+   to do this currently.
+
+   With this proposal, we can do this::
+
+     -- top of file:
+     module MyLibrary ( module C(meth) ) where
+
+       class C a where ...
+
+       import module C where
+         meth :: ...
+
+   This example shows that modules may be *extended*. The ``class C``
+   declaration implicitly creates module ``C``, which is then extended below.
+
+   An import specifier of ``C(..)`` will not import ``meth`` after this
+   change. Instead, importers must say ``import module C``. However, this new
+   import statement is a drop-in replacement for ``C(..)`` and may become
+   preferable (as it is customizable in the way demonstrated here). Thus,
+   it is reasonable that Haskellers would learn to write ``import module C``
+   in import lists instead of ``C(..)``, as the former (new form) is
+   extensible.
```

Comment:
Maybe this is answered elsewhere in the proposal, but why is that `C(..)` should not import `meth`?

Reply:
Hm, answering myself, `C(..)` refers to the class, not to the module that happens to be named the same.

Reply:
Yep -- you figured it out.